src/main/java/com/linkedin/kmf/services/ClusterTopicManipulationService.java

```diff
@@ -0,0 +1,248 @@
+  private final ScheduledExecutorService _executor;
+  private final AdminClient _adminClient;
+  private volatile boolean _isOngoingTopicCreationDone;
+  private final AtomicBoolean _running;
+  private String _currentlyOngoingTopic;
+  private CreateTopicsResult _createTopicsResult;
+  private AtomicInteger _totalPartitions = new AtomicInteger();
+  // TODO -- ClusterTopicManipulationMetrics implementation in progress!
+  private ClusterTopicManipulationMetrics _clusterTopicManipulationMetrics;
+
+  public ClusterTopicManipulationService(String name, AdminClient adminClient) {
+    LOGGER.info("ClusterTopicManipulationService constructor initiated {}", this.getClass().getName());
+
+    _isOngoingTopicCreationDone = true;
+    _adminClient = adminClient;
+    _executor = Executors.newSingleThreadScheduledExecutor();
+    _reportIntervalSecond = Duration.ofSeconds(1);
+    _running = new AtomicBoolean(false);
+    _configDefinedServiceName = name;
+    // TODO: instantiate a new instance of ClusterTopicManipulationMetrics(..) here.
+  }
+
+  /**
+   * The start logic must only execute once.  If an error occurs then the implementer of this class must assume that
+   * stop() will be called to clean up.  This method must be thread safe and must assume that stop() may be called
+   * concurrently. This can happen if the monitoring application's life cycle is being managed by a container.  Start
+   * will only be called once.
+   */
+  @Override
+  public void start() {
+    if (_running.compareAndSet(false, true)) {
+      LOGGER.info("ClusterTopicManipulationService started for {} - {}", _configDefinedServiceName,
+          this.getClass().getCanonicalName());
+      Runnable clusterTopicManipulationServiceRunnable = new ClusterTopicManipulationServiceRunnable();
+
+      _executor.scheduleAtFixedRate(clusterTopicManipulationServiceRunnable, _reportIntervalSecond.getSeconds(),
+          _reportIntervalSecond.getSeconds(), TimeUnit.SECONDS);
+    }
+  }
+
+  private class ClusterTopicManipulationServiceRunnable implements Runnable {
+
+    private ClusterTopicManipulationServiceRunnable() {
+      // unaccessed.
+    }
+
+    /**
+     * When an object implementing interface <code>Runnable</code> is used
+     * to create a thread, starting the thread causes the object's
+     * <code>run</code> method to be called in that separately executing
+     * thread.
+     * <p>
+     * The general contract of the method <code>run</code> is that it may
+     * take any action whatsoever.
+     *
+     * @see     Thread#run()
+     */
+    @Override
+    public void run() {
+      try {
+        ClusterTopicManipulationService.this.createDeleteClusterTopic();
+      } catch (Exception e) {
+        LOGGER.error("{} {} failed to run createDeleteClusterTopic()", _configDefinedServiceName,
+            ClusterTopicManipulationService.this.getClass().getSimpleName(), e);
+      }
+    }
+  }
+
+  private void createDeleteClusterTopic() {
```

Comment:
This method is probably the most important part of this PR (or this whole feature). Several things:
1. Please add Java doc to explain steps this method is taking. It took me a while to understand. In the Java doc, explicitly state whether this method is thread-safe or does it need to handle concurrent accesses.

2. IIUC, this method is thread-safe since you are using a `newSingleThreadScheduledExecutor` to schedule it. If that's true, what's the point of using a volatile boolean `_isOngoingTopicCreationDone` given that only one thread can access this method at a time?

3.  It seems that we are not waiting for topics to be created. Looking at `if (_createTopicsResult.all().isDone() && doesClusterContainTopic(_currentlyOngoingTopic, brokers,
          _adminClient))`, I think if a topic does not exist at this point, the `else` branch is executed. I thought we want to measure how long it takes to create a topic right? If there is no blocking and waiting for a topic to be created, how do we know how long it takes to create the topic?

Maybe my above questions are results of myself not understanding exactly what this method is doing. So, please clarify and add Java doc

Reply:
Added the JavaDoc.
3 - there is no else branch. If the topic partitions log directories aren't fully populated at that point, it will exit `createDeleteClusterTopic` method and in the next iteration of the `ClusterTopicManipulationServiceRunnable`, it will check again if the topic has been created across the brokers. 
I understand your point that we could just block. But the ClusterTopicManipulationServiceRunnable repeats and runs it self periodically at a non-trivial interval anyway to measure the latency at a granular level.

Reply:
I kind of understand your point. I think it's like `polling` vs `interrupt`. If we block and wait on the KafkaFuture, it's like `interrupt` whereas your approach is like `polling` (keep asking if something has happened).

I still think it's very awkward to use this `polling` method for 2 reasons.
1. We already have the opportunity to block and wait since we can wait on the KafkaFuture. In other words, Kafka topic creation/deletion APIs can let us block and wait. Why are we still going with the less accurate and less intuitive polling method? 

2. As you said `ClusterTopicManipulationServiceRunnable  repeats and runs itself periodically at a non-trivial interval`, that means the accuracy of the measure depends on the frequency at which the ClusterTopicManipulationServiceRunnable calls the method `createDeleteClusterTopic`. Whereas if we block and wait on the KafkaFuture, it is very accurate regarding how much time it took for the controller to create/delete a topic

Reply:
I have no disagreement nor a large affinity toward one solution.
If you think that the level of granularity needs to be more sensitive than it currently is, it's perfectly fine calling `.all().get()` on the KafkaFuture to block the call.
I will make that change.

Reply:
Sounds good. My whole point is that `calling .all().get() on the KafkaFuture to block the call` is easily available to us and will give us the most accurate measure. Plus it is how users of the AdminClient check whether their topics are created or deleted successfully. 

So There is no reason for not taking it.

Reply:
Okay to resolve this?

Reply:
OK

Reply:
As for feasibility of having both of these:
1 - fetching the metadata or doing describeTopic and
2 - verifying that the log dirs have the topic partitions

All Kafka brokers can answer a topic metadata request that describes the current state of the cluster, such as what topics there are, which partitions those topics have, which broker is the leader for those partitions, and the host and port information for these brokers.
The client finds a broker and that broker tells the client regarding all the other brokers which exist and what partitions they host. (This first broker may itself go down so the best practice for a client is to take ask two or three bootstrap brokers).
The topic creation could be issued through AdminClient and we could have a scenario where we send a request to broker A for metadata fetch. The broker could fetch the metadata and return a good signal and but in reality the broker B could be experiencing extremely slow speed/network and so it might not have even finished creating these topic partitions (propagating metadata).
At the point in time, checking the log directories of broker B will return negative (bad) since the topic partitions haven’t been created for that partition.